src/Compilers/CSharp/Portable/Binder/Binder_Await.cs

```diff
@@ -115,8 +118,34 @@ private bool CouldBeAwaited(BoundExpression expression)
                 return false;
             }
 
+            if (containingMethod is SynthesizedSimpleProgramEntryPointSymbol { IsAsync: false })
+            {
+                var bag = BindingDiagnosticBag.GetInstance(withDependencies: false, withDiagnostics: true);
+                try
+                {
+                    _ = BindAwait(expression, expression.Syntax, bag);
+                    // Top-level statements are always allowed to use `await`, so we
+                    // can ignore specific errors
+                    Debug.Assert(bag.DiagnosticBag is not null);
+                    return !bag.DiagnosticBag.AsEnumerableWithoutResolution().Any(isErrorOtherThanBadAwait);
+                }
+                finally
+                {
+                    bag.Free();
+                }
+            }
+
             var boundAwait = BindAwait(expression, expression.Syntax, BindingDiagnosticBag.Discarded);
             return !boundAwait.HasAnyErrors;
+
+            static bool isErrorOtherThanBadAwait(Diagnostic diagnostic)
```

Comment:
It is probably better (from readability perspective) to convert this into a lambda at the only use site.